wsd/ProxyRequestHandler.cpp

```diff
@@ -0,0 +1,99 @@
+ */
+
+#include <config.h>
+
+#include <Poco/URI.h>
+
+#include <COOLWSD.hpp>
+#include "ProxyRequestHandler.hpp"
+#include <net/HttpRequest.hpp>
+#include <net/HttpHelper.hpp>
+
+std::map<unsigned long, std::shared_ptr<StreamSocket>> ProxyRequestHandler::LinkSession;
+std::map<std::string, std::shared_ptr<http::Response>> ProxyRequestHandler::CacheFileHash;
+std::chrono::system_clock::time_point ProxyRequestHandler::MaxAge;
+
+void ProxyRequestHandler::handleRequest(const std::string& relPath,
+                                        const std::shared_ptr<StreamSocket>& socket)
+{
+    Poco::URI uriProxy(ProxyServer);
+    const auto timeNow = std::chrono::system_clock::now();
+
+    if (MaxAge > std::chrono::system_clock::time_point() && timeNow > MaxAge)
+    {
+        CacheFileHash.clear();
+        MaxAge = std::chrono::system_clock::time_point();
+    }
+
+    const auto cacheEntry = CacheFileHash.find(relPath);
+    if (cacheEntry != CacheFileHash.end())
+    {
+        socket->sendAndShutdown(*cacheEntry->second.get());
+        return;
+    }
+
+    uriProxy.setPath(relPath);
+    auto sessionProxy = http::Session::create(uriProxy.getHost(),
+                                              http::Session::Protocol::HttpSsl,
+                                              uriProxy.getPort());
+    sessionProxy->setTimeout(std::chrono::seconds(10));
+    http::Request requestProxy(uriProxy.getPathAndQuery());
+    http::Session::FinishedCallback proxyCallback =
+        [](const std::shared_ptr<http::Session>& httpSession)
+            {
+                const auto callbackNow = std::chrono::system_clock::now();
+                auto sessionLink = LinkSession.find(reinterpret_cast<unsigned long>(httpSession.get()));
+                if (sessionLink == LinkSession.end())
+                {
+                    return;
+                }
+
+                try
+                {
+                    const std::shared_ptr<const http::Response> httpResponse = httpSession->response();
+                    if (httpResponse->statusLine().statusCode() == 200)
+                    {
+                        if (MaxAge == std::chrono::system_clock::time_point())
+                        {
+                            MaxAge = callbackNow + std::chrono::hours(10);
+                        }
+
+                        std::shared_ptr<http::Response> entry =
+                            std::make_shared<http::Response>(*httpResponse.get());
+                        CacheFileHash[httpSession->getUrl()] = entry;
+
+                        entry->add("Server", HTTP_SERVER_STRING);
+                        entry->add("Date", Util::getHttpTimeNow());
+                        sessionLink->second->sendAndShutdown(*entry.get());
+                    }
+                    else
+                    {
+                        HttpHelper::sendErrorAndShutdown(400, sessionLink->second);
+                    }
+                }
+                catch(...)
```

Comment:
Seems useful to catch `std::exception`, which more often than not will have valuable information for us. Also, reasonable to log as error here, or possibly warning, since these should be rare, but useful to understand edge-cases better.

Reply:
fair enough, I did it, but it seems forget to stage the chunk